### PR TITLE
ApiManagement - namedValue secrets and createContracts with required properties

### PR DESCRIPTION
**Pull request by vfedonkin**
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/adx-documentation-pr/wiki/Overall-basic-flow) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

### CODE REVIEW

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json [hunk 1]
```diff
@@ -2830,6 +2869,53 @@
       },
       "description": "Paged OpenIdProviders list representation."
     },
+    "OpenidConnectProviderCreateContract": {
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json]: >OpenidConnectProviderCreateContract [](start = 5, length = 35)

maybe we can have `OpenIDConnectProvider` 
[https://auth0.com/docs/protocols/oidc](https://auth0.com/docs/protocols/oidc)

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json [hunk 2]
```diff
@@ -2830,6 +2869,53 @@
...
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/OpenidConnectProviderContractProperties",
+          "description": "OpenId Connect Provider contract properties."
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json]: >OpenId [](start = 26, length = 6)

rename to `OpenID`

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json [hunk 3]
```diff
@@ -2830,6 +2869,53 @@
...
+        }
+      },
+      "allOf": [
+        {
+          "$ref": "./apimanagement.json#/definitions/Resource"
+        }
+      ],
+      "description": "OpenId Connect Provider details."
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json]: >Connect [](start = 29, length = 7)

create contract

**Comment by solankisamir** [comment]
>         }

we dont return `clientSecret` in GET... then why have it

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json:2956 in 544e8d3. [](commit_id = 544e8d33af37343a60f17b0cf821eb0c8db859e4, deletion_comment = False)

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json [hunk 4]
```diff
@@ -5189,6 +5275,15 @@
...
       "description": "Client or app secret used in IdentityProviders, Aad, OpenID or OAuth."
     },
+    "PropertyValueContract": {
+      "properties": {
+        "value": {
+          "type": "string",
+          "description": "Secret Value of the NamedValue entity."
+        }
+      },
+      "description": "Client or app secret used in IdentityProviders, Aad, OpenID or OAuth."
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json]: >Client or app secret used in IdentityProviders, Aad, OpenID or OAuth. [](start = 22, length = 69)

fix description. This is secret value of the NamedValue entity.

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json (hunk 2 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json]: Done

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json (hunk 3 above)
Comment (vfedonkin) [reply specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json]: Done

**Comment by vfedonkin** [comment]
@solankisamir , ClientSecret removed from Get. Could you please look again?

**Comment by majastrz** [comment]
>         "clientSecret": {

Can you add x-ms-secret: true to mark it as a secret?

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json:2277 in 1535884. [](commit_id = 1535884bbeebf583cc597ecc473f5919d2114440, deletion_comment = False)

**Comment by majastrz** [comment]
>         "clientSecret": {

Please do this on any property that accepts a secret value.

---
In reply to: [584879568](https://github.com/Azure/azure-rest-api-specs/pull/8341#issuecomment-584879568) [](ancestors = 584879568)

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json:2277 in 1535884. [](commit_id = 1535884bbeebf583cc597ecc473f5919d2114440, deletion_comment = False)

**Review by majastrz — CHANGES_REQUESTED** [review]
Can you add the x-ms-secret: true annotation to designate secret properties. I'll be able to sign off from ARM side after you make the change.

**Review by majastrz — APPROVED** [review]
Signed off from ARM side.

**Comment by solankisamir** [comment]
>         }

why does this contract have `value`

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/definitions.json:3713 in 3ffd5be. [](commit_id = 3ffd5bebe959ea88cce759bb3908e5a7db02d251, deletion_comment = False)

**Comment by solankisamir** [comment]
>               "$ref": "./definitions.json#/definitions/NamedValueContract"

we should have a `createcontract` for this entity too.

---
Refers to: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/apimnamedvalues.json:221 in 3ffd5be. [](commit_id = 3ffd5bebe959ea88cce759bb3908e5a7db02d251, deletion_comment = False)

**Comment by vfedonkin** [comment]
@solankisamir , NamedValues don't always contains the secrets. We return 'value' property on GET and LIST in case "isSecret" property == false

**Comment by solankisamir** [comment]
@vitaliy.fedonkin@gmail.com  I think we should have consistent experience, so that customer doesn't have to think whether GET will return him secrets or not.

---
In reply to: [584976201](https://github.com/Azure/azure-rest-api-specs/pull/8341#issuecomment-584976201) [](ancestors = 584976201)

**Comment by vfedonkin** [comment]
@solankisamir , you mean removing "isSecret" property and consider all NamedValues as secrets? Yep, I will do it.

**Comment by erich-wang** [comment]
@solankisamir, @vfedonkin , please let me know when you have agreement and all concerns have been addressed, then I can merge this PR, thanks.

**Comment by vfedonkin** [comment]
@solankisamir , namedValues are changed. Could you please validate?

File: specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/examples/ApiManagementUpdateNamedValue.json [hunk 5]
```diff
@@ -26,13 +28,12 @@
...
-        "name": "testprop2",
         "properties": {
           "displayName": "prop3name",
           "value": "propValue",
           "tags": [
             "foo",
-            "bar"
+            "bar2"
           ],
           "secret": true
```
Comment (solankisamir) [inline specification/apimanagement/resource-manager/Microsoft.ApiManagement/preview/2019-12-01-preview/examples/ApiManagementUpdateNamedValue.json]: >true [](start = 20, length = 4)

we wont change the secretvalue from `false` to `true` :)

**Review by solankisamir — APPROVED** [review]
:shipit:

**Comment by vfedonkin** [comment]
@erich-wang , the PR is ready for merge, thank you.

**Pull request merged**